src/flag_gems/ops/scan.py

```diff
@@ -0,0 +1,78 @@
+import logging
```

Comment:
The functions in this file are better building blocks for cumsum. Shall we pack and move these functions into cumsum.py?

Reply:
Is there any plan to develop scan operator? If there is, then I think a separate scan operator is better.

Reply:
There's no corresponding scan operator in aten, is there?

Reply:
No, there is no corresponding scan operator in aten.
There is a 'ScanKernels' in native cuda which contains the implementation for cumsum, cumprod and ect. Could we also treat scan just like this 'ScanKernels' not a operator?

Reply:
add a ABC version scan_then_fan in scan, which can support M and K are not 1.
perf on A100, batch=1024:
```
Operator cumsum Performance Test (torch.float16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.077824            0.014336
6144                   0.15872            0.078848
11264                 0.282624            0.089088
16384                 0.406528            0.131072
21504                 0.370688            0.166912
26624                 0.456704            0.200704
31744                  0.54272             0.23552
36864                 0.628736            0.272384
41984                 0.713728              0.3072
47104                 0.800768            0.345088
52224                  0.88576            0.375808
57344                   0.9728             0.41472
62464                  1.05779            0.454656
67584                 0.999424            0.482304
72704                  1.07622            0.524288
77824                  1.14995             0.55808
Operator cumsum Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.080896            0.014336
6144                  0.147456              0.0768
11264                 0.263168            0.137216
16384                 0.376832             0.18944
21504                  0.33792            0.244736
26624                 0.415744            0.299008
31744                 0.493568            0.352256
36864                 0.571392              0.4096
41984                 0.649216            0.463872
47104                  0.72704            0.514048
52224                 0.804864            0.566272
57344                 0.883712            0.627712
62464                 0.961536            0.684032
67584                 0.945152             0.73216
72704                  1.01376            0.821248
77824                  1.08339            0.848896
Operator cumsum Performance Test (torch.bfloat16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.077824            0.013312
6144                  0.159744            0.079872
11264                 0.284672            0.089088
16384                   0.4096            0.130048
21504                 0.380928            0.166912
26624                 0.467968            0.200704
31744                 0.557056            0.236544
36864                  0.64512             0.27136
41984                 0.733184            0.306176
47104                 0.821248            0.344064
52224                 0.909312            0.376832
57344                   0.9984             0.41472
62464                  1.08544            0.454656
67584                  1.02707            0.482304
72704                  1.10387             0.52224
77824                  1.17965            0.553984
```

Reply:
The triton flavored scan operator should support combine functions other than add. Triton already has its block level builtin baked in. You are free to provide a general scan operator. For now let's just replace original cumsum. Shall we?

Reply:
> The triton flavored scan operator should support combine functions other than add. Triton already has its block level builtin baked in. You are free to provide a general scan operator. For now let's just replace original cumsum. Shall we?

Sure, done.

Reply:
> > > The triton flavored scan operator should support combine functions other than add. Triton already has its block level builtin baked in. You are free to provide a general scan operator. For now let's just replace original cumsum. Shall we?
> > 
> > 
> > Sure, done.
> 
> How does the time it takes to autotune the scan op compare to the time it takes to autotune the original cumsum op?

plz see the comment below

Reply:
> add a ABC version scan_then_fan in scan, which can support M and K are not 1. perf on A100, batch=1024:
> 
> ```
> Operator cumsum Performance Test (torch.float16)
> Size        Torch Latency (ms)   Gems Latency (ms)
> --------------------------------------------------
> 1024                  0.077824            0.014336
> 6144                   0.15872            0.078848
> 11264                 0.282624            0.089088
> 16384                 0.406528            0.131072
> 21504                 0.370688            0.166912
> 26624                 0.456704            0.200704
> 31744                  0.54272             0.23552
> 36864                 0.628736            0.272384
> 41984                 0.713728              0.3072
> 47104                 0.800768            0.345088
> 52224                  0.88576            0.375808
> 57344                   0.9728             0.41472
> 62464                  1.05779            0.454656
> 67584                 0.999424            0.482304
> 72704                  1.07622            0.524288
> 77824                  1.14995             0.55808
> Operator cumsum Performance Test (torch.float32)
> Size        Torch Latency (ms)   Gems Latency (ms)
> --------------------------------------------------
> 1024                  0.080896            0.014336
> 6144                  0.147456              0.0768
> 11264                 0.263168            0.137216
> 16384                 0.376832             0.18944
> 21504                  0.33792            0.244736
> 26624                 0.415744            0.299008
> 31744                 0.493568            0.352256
> 36864                 0.571392              0.4096
> 41984                 0.649216            0.463872
> 47104                  0.72704            0.514048
> 52224                 0.804864            0.566272
> 57344                 0.883712            0.627712
> 62464                 0.961536            0.684032
> 67584                 0.945152             0.73216
> 72704                  1.01376            0.821248
> 77824                  1.08339            0.848896
> Operator cumsum Performance Test (torch.bfloat16)
> Size        Torch Latency (ms)   Gems Latency (ms)
> --------------------------------------------------
> 1024                  0.077824            0.013312
> 6144                  0.159744            0.079872
> 11264                 0.284672            0.089088
> 16384                   0.4096            0.130048
> 21504                 0.380928            0.166912
> 26624                 0.467968            0.200704
> 31744                 0.557056            0.236544
> 36864                  0.64512             0.27136
> 41984                 0.733184            0.306176
> 47104                 0.821248            0.344064
> 52224                 0.909312            0.376832
> 57344                   0.9984             0.41472
> 62464                  1.08544            0.454656
> 67584                  1.02707            0.482304
> 72704                  1.10387             0.52224
> 77824                  1.17965            0.553984
> ```

I ran the cumsum benchmark on A100, the result of float32 is different:
test_reduction_perf.py Operator cumsum Performance Test (torch.float16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.074944            0.012896
6144                  0.124256             0.04704
11264                 0.220736            0.079232
16384                 0.315392            0.125248
21504                 0.284992            0.159616
26624                  0.35088            0.191744
31744                 0.416768            0.225632
36864                 0.482048            0.260576
41984                 0.548352            0.293824
47104                 0.613568            0.337056
52224                 0.678784            0.359104
57344                 0.746112            0.401952
62464                   0.8104            0.442272
67584                 0.771552            0.464224
72704                 0.828736            0.508992
77824                 0.885536            0.537536
Operator cumsum Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                   0.06416            0.013664
6144                  0.117504            0.070144
11264                 0.207968            0.150912
16384                 0.295808            0.213248
21504                 0.263392            0.274528
26624                  0.32352            0.336928
31744                 0.383744            0.398848
36864                 0.444832            0.470752
41984                 0.504832              0.5312
47104                 0.565056            0.590976
52224                 0.624896            0.649152
57344                 0.684736            0.723904
62464                 0.745184            0.789696
67584                 0.741888            0.844032
72704                 0.796416            0.952448
77824                 0.850624             0.98816
Operator cumsum Performance Test (torch.bfloat16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.061248              0.0128
6144                  0.125184            0.047104
11264                 0.221408            0.078752
16384                  0.31744            0.124096
21504                  0.29152             0.15824
26624                 0.359296            0.191296
31744                 0.426912             0.22592
36864                 0.494432            0.260224
41984                 0.561792             0.29264
47104                 0.628352            0.333312
52224                 0.695808            0.360288
57344                 0.763744            0.407584
62464                 0.829856            0.449312
67584                   0.7904            0.464032
72704                 0.849376            0.514336
77824                 0.907168            0.539552
.

Reply:
I switched to another A100 and got a torch performance close to yours. The Gems's perf alse become faster, which is close to the perf of torch in big case.
```
Operator cumsum Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.063488            0.012288
6144                  0.115712            0.083968
11264                   0.2048            0.131072
16384                 0.293888            0.183296
21504                 0.260096            0.237568
26624                 0.320512             0.29184
31744                 0.380928            0.344064
36864                  0.44032            0.401408
41984                 0.499712             0.45568
47104                 0.559104            0.504832
52224                 0.618496             0.55808
57344                 0.678912            0.621568
62464                 0.738304            0.679936
67584                 0.731136            0.720896
72704                 0.786432            0.787456
77824                  0.83968            0.836608
```